pkg/envoy/xds/cache.go

```diff
@@ -105,27 +109,49 @@ func (c *Cache) tx(typeURL string, upsertedResources map[string]proto.Message, d
 		// responses in GetResources.
 		// Calling proto.Message.String is not very cheap, but we assume that
 		// the reduced churn between the clients and the server is worth it.
-		if force || !found || oldV.resource.String() != value.String() {
+		if !found || oldV.resource.String() != value.String() {
 			if found {
 				cacheLog.WithField(logfields.XDSResourceName, name).Debug("updating resource in cache")
+
+				if revertUpsertedResources == nil {
+					revertUpsertedResources = make(map[string]proto.Message, len(upsertedResources)+len(deletedNames))
+				}
+				revertUpsertedResources[name] = oldV.resource
 			} else {
 				cacheLog.WithField(logfields.XDSResourceName, name).Debug("inserting resource into cache")
+
+				if revertDeletedNames == nil {
+					revertDeletedNames = make([]string, 0, len(upsertedResources))
+				}
+				revertDeletedNames = append(revertDeletedNames, name)
```

Comment:
And this one `deleteInsertedNames`?

Reply:
Likewise, this slice will be passed as the `deletedNames` parameter, and I wanted to match that name.